sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala

```diff
@@ -61,7 +61,11 @@ case class BoundReference(ordinal: Int, dataType: DataType, nullable: Boolean)
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     if (ctx.currentVars != null && ctx.currentVars(ordinal) != null) {
       val oev = ctx.currentVars(ordinal)
-      ev.isNull = oev.isNull
+      if (nullable) {
```

Comment:
I think this kind of optimizations are already done in https://github.com/apache/spark/pull/19813/files#diff-b3ebf3b40b9d4b6e98bb29ac8bb5aadaR108

Reply:
oh, right.